Start moving classes which use SharedBitmap across to the new Serialization format.

gavin-apple opened the pull request:
#### 62d66343552b42ae416f047573690a0a61dcdc30
<pre>
Start moving classes which use SharedBitmap across to the new Serialization format.
<a href="https://bugs.webkit.org/show_bug.cgi?id=246087">https://bugs.webkit.org/show_bug.cgi?id=246087</a>
&lt;rdar://100808518&gt;

Reviewed by Alex Christensen.

This change starts to move some of the classes which use SharedBitmap
across to the new Serialization format. This includes:
     * InteractionInformationAtPosition
     * InteractionInformationAtPosition
     * Selectability
     * PasteboardWebContent
     * PasteboardBuffer
     * PasteboardURL
     * PasteboardImage

It also adds the ability to have conditional enum declarations.

* Source/WebCore/platform/Cursor.h:
(WebCore::Cursor::image const):
* Source/WebCore/platform/Pasteboard.cpp:
* Source/WebCore/platform/Pasteboard.h:
* Source/WebCore/platform/cocoa/PasteboardCocoa.mm:
* Source/WebKit/CMakeLists.txt:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Scripts/generate-serializers.py:
(SerializedEnum.__init__):
(generate_header):
(generate_impl):
(parse_serialized_types):
* Source/WebKit/Shared/ShareableBitmap.cpp:
(WebKit::ShareableBitmap::createReadOnly):
(WebKit::ShareableBitmap::createReadOnlyHandle const):
* Source/WebKit/Shared/ShareableBitmap.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.cpp:
(IPC::ArgumentCoder&lt;RefPtr&lt;Image&gt;&gt;::encode):
(IPC::ArgumentCoder&lt;RefPtr&lt;Image&gt;&gt;::decode):
(IPC::encodeOptionalImage):
(IPC::decodeOptionalImage):
(IPC::ArgumentCoder&lt;Cursor&gt;::encode):
(IPC::ArgumentCoder&lt;Cursor&gt;::decode):
(IPC::ArgumentCoder&lt;TextIndicatorData&gt;::encode):
(IPC::ArgumentCoder&lt;WebCore::NotificationResources&gt;::encode):
(IPC::encodeImage): Deleted.
(IPC::decodeImage): Deleted.
(IPC::ArgumentCoder&lt;PasteboardWebContent&gt;::encode): Deleted.
(IPC::ArgumentCoder&lt;PasteboardWebContent&gt;::decode): Deleted.
(IPC::ArgumentCoder&lt;PasteboardImage&gt;::encode): Deleted.
(IPC::ArgumentCoder&lt;PasteboardImage&gt;::decode): Deleted.
* Source/WebKit/Shared/WebCoreArgumentCoders.h:
* Source/WebKit/Shared/ios/InteractionInformationAtPosition.h:
(WebKit::InteractionInformationAtPosition::isSelectable const):
* Source/WebKit/Shared/ios/InteractionInformationAtPosition.mm:
(WebKit::InteractionInformationAtPosition::encode const): Deleted.
(WebKit::InteractionInformationAtPosition::decode): Deleted.
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView textInteractionGesture:shouldBeginAtPoint:]):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::selectionPositionInformation):

Canonical link: <a href="https://commits.webkit.org/255513@main">https://commits.webkit.org/255513@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/aabb875f6d596cddd5a91beb821573b4245bb38d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/92712 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/1926 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/23298 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/102428 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/162790 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/96714 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/1926 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/30277 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/85099 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/98576 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/98375 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/1299 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/79204 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/28236 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/5/builds/96288 "Passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/83213 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/82924 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/71319 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/36682 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/16844 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/34474 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/18031 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3818 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/38346 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/40638 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/40257 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/37199 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebKit/Scripts/generate-serializers.py:
```diff
@@ -66,6 +66,7 @@ class SerializedEnum(object):
     def __init__(self, namespace, name, underlying_type, valid_values, condition, attributes):
         self.namespace = namespace
         self.name = name
+        self.condition = condition
```
This is already done on line 72.

achristensen07 commented on Source/WebKit/Shared/ios/InteractionInformationAtPosition.h:
```diff
@@ -41,6 +41,14 @@
 
 namespace WebKit {
 
+enum class Selectability : uint8_t {
```
You will likely not need this change if you use the [Nested] attribute in the .serialization.in

achristensen07 commented on Source/WebKit/Shared/ios/InteractionInformationAtPosition.serialization.in:
```diff
@@ -0,0 +1,96 @@
+    UnselectableDueToMediaControls,
+};
+
+[CustomHeader=True] struct WebKit::InteractionInformationAtPosition {
```
This shouldn't need [CustomHeader]

achristensen07 commented on Source/WebKit/Scripts/generate-serializers.py:
```diff
@@ -329,12 +329,25 @@ def generate_impl(serialized_types, serialized_enums, headers):
-                if 'ReturnEarlyIfTrue' in member.attributes:
-                    result.append('    if (*' + sanitize_string_for_variable_name(member.name) + ')')
-                    result.append('        return { ' + type.namespace_and_name() + ' { } };')
+                r = re.compile("SecureCodingAllowed=\[(.*)\]")
```
When we add new things to this script, I like to add to Source/WebKit/Scripts/webkit/tests/TestSerializedType.serialization.in then run "make -C Source/WebKit/Scripts/webkit/tests" so we have a "test" of what the generated code is supposed to look like.

achristensen07 commented on Source/WebKit/Shared/Pasteboard.serialization.in:
```diff
@@ -0,0 +1,90 @@
+
+headers: <WebCore/Pasteboard.h>
+
+header: <WebCore/Pasteboard.h>
```
Also, this is redundant.

gavin-apple committed:
Start moving classes which use SharedBitmap across to the new Serialization format.
https://bugs.webkit.org/show_bug.cgi?id=246087
<rdar://100808518>

Reviewed by Alex Christensen.

This change starts to move some of the classes which use SharedBitmap
across to the new Serialization format. This includes:
     * InteractionInformationAtPosition
     * InteractionInformationAtPosition
     * Selectability
     * PasteboardWebContent
     * PasteboardBuffer
     * PasteboardURL
     * PasteboardImage

It also adds the ability to have conditional enum declarations.

* Source/WebCore/platform/Cursor.h:
(WebCore::Cursor::image const):
* Source/WebCore/platform/Pasteboard.cpp:
* Source/WebCore/platform/Pasteboard.h:
* Source/WebCore/platform/cocoa/PasteboardCocoa.mm:
* Source/WebKit/CMakeLists.txt:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Scripts/generate-serializers.py:
(SerializedEnum.__init__):
(generate_header):
(generate_impl):
(parse_serialized_types):
* Source/WebKit/Shared/ShareableBitmap.cpp:
(WebKit::ShareableBitmap::createReadOnly):
(WebKit::ShareableBitmap::createReadOnlyHandle const):
* Source/WebKit/Shared/ShareableBitmap.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.cpp:
(IPC::ArgumentCoder<RefPtr<Image>>::encode):
(IPC::ArgumentCoder<RefPtr<Image>>::decode):
(IPC::encodeOptionalImage):
(IPC::decodeOptionalImage):
(IPC::ArgumentCoder<Cursor>::encode):
(IPC::ArgumentCoder<Cursor>::decode):
(IPC::ArgumentCoder<TextIndicatorData>::encode):
(IPC::ArgumentCoder<WebCore::NotificationResources>::encode):
(IPC::encodeImage): Deleted.
(IPC::decodeImage): Deleted.
(IPC::ArgumentCoder<PasteboardWebContent>::encode): Deleted.
(IPC::ArgumentCoder<PasteboardWebContent>::decode): Deleted.
(IPC::ArgumentCoder<PasteboardImage>::encode): Deleted.
(IPC::ArgumentCoder<PasteboardImage>::decode): Deleted.
* Source/WebKit/Shared/WebCoreArgumentCoders.h:
* Source/WebKit/Shared/ios/InteractionInformationAtPosition.h:
(WebKit::InteractionInformationAtPosition::isSelectable const):
* Source/WebKit/Shared/ios/InteractionInformationAtPosition.mm:
(WebKit::InteractionInformationAtPosition::encode const): Deleted.
(WebKit::InteractionInformationAtPosition::decode): Deleted.
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView textInteractionGesture:shouldBeginAtPoint:]):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::selectionPositionInformation):

Canonical link: https://commits.webkit.org/255513@main
